Restore ansible --version output

samdoran opened the pull request:
##### SUMMARY

Fixes #55710

The default formatting class used by `argparse` strips newline characters, resulting in the output of `ansible --version` being a single string.

Implement a custom action to print out the already formatted version string and add tests to prevent future regressions.

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request

##### COMPONENT NAME
`/lib/ansible/cli/arguments/option_helpers.py`

##### ADDITIONAL INFORMATION
This only affects 2.9.

samdoran committed:
Add custom action class for version info

samdoran committed:
Add integration test for checking ansible --version

AlanCoding commented on test/integration/targets/ansible/test_version.py:
```diff
@@ -0,0 +1,17 @@
+version_lines = version_string.splitlines()
+
+assert len(version_lines) == 6, 'Incorrect number of lines in "ansible --version" output'
+assert re.match(b'ansible [0-9.a-z]+$', version_lines[0]), 'Incorrect ansible version line in "ansible --version" output'
```
👍 this should cover it, thanks!

sivel commented on test/integration/targets/ansible/test_version.py:
```diff
@@ -0,0 +1,17 @@
+#!/usr/bin/env python
```
Instead of using a script that subprocesses ansible, you can add a test to `test/units/cli/test_adhoc.py` that utilizes the `capsys` pytest fixture to get the output: https://docs.pytest.org/en/latest/capture.html

samdoran replied:
Nice! I will check that out.

samdoran replied:
I have this mostly working. The only issue I have when running using `AdHocCLI` is the `%(prog)s` is not `ansible` but a `pytest` string:

```
test/units/cli/test_adhoc.py pytest 2.9.0.dev0
  config file = /Users/sdoran/.ansible.cfg
  configured module search path = ['/Users/sdoran/Source/ansible/library']
  ansible python module location = /Users/sdoran/Source/ansible/lib/ansible
  executable location = /Users/sdoran/.pyenv/versions/3.7.3/bin/pytest
  python version = 3.7.3 (default, Apr 18 2019, 13:51:13) [Clang 10.0.1 (clang-1001.0.46.3)]
```

Is there any way to fix that? Because I think testing the content of that first line is the most critical part.

sivel replied:
There are likely a few ways to address that.  

Looking at how you are calling the `version` function:

```
ansible_version = to_native(version(getattr(parser, 'prog')))
```

That is using `prog` from `parser`.  `parser` is available at `adhoc_cli.parser` so you can likely just monkeypatch that.

Otherwise, maybe monkeypatch `sys.argv[0]`.

samdoran replied:
I wasn't having any luck patching the `parser` object, but patching `sys.argv` did work.

samdoran committed:
Change to unit test using capsys

abadger commented:
I think this is what you need: https://docs.python.org/2/library/argparse.html#prog

prog should be set to: ```os.path.basename(self.args[0])```.  you'll need to pass that from the Cli object to create_base_parser() as a new function parameter.

AlanCoding commented:
I pulled this branch and got:

```
(ansible3) arominge-OSX:ansible alancoding$ ansible --version
ansible 2.9.0.dev0
  config file = None
  configured module search path = ['/Users/alancoding/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/alancoding/Documents/repos/ansible/lib/ansible
  executable location = /Users/alancoding/.virtualenvs/ansible3/bin/ansible
  python version = 3.6.5 (default, Apr 25 2018, 14:23:58) [GCC 4.2.1 Compatible Apple LLVM 9.1.0 (clang-902.0.39.1)]
```

Looks good!

samdoran committed:
Improvements to unit test

- Fix capsys error on Python 2.6
- Patch sys.argv in order to get the correct value in version output

samdoran committed:
Use args from CLI as prog for ArgumentParser object

Update unit tests

samdoran commented on lib/ansible/cli/__init__.py:
```diff
@@ -275,7 +275,7 @@ def init_parser(self):
                 self.parser.add_option('--my-option', dest='my_option', action='store')
         """
-        self.parser = opt_help.create_base_parser(usage=usage, desc=desc, epilog=epilog)
+        self.parser = opt_help.create_base_parser(usage=usage, desc=desc, epilog=epilog, prog=os.path.basename(self.args[0]))
```
@abadger I'm not sure if this is a good default. If `args` is an empty list (as it was in the unit test), this raises an `IndexError`. Should I do some sanity checking or `args` or will that never be the case during normal operation?

abadger replied:
During normal operation, it would never be empty (A program is always invoked with at least its filename as an argument.)

Perhaps we should sanity check in the ```CLI``` constructor.

``` python
def __init__(self, args, callback=None):
    if not args:
        raise ValueError('Require a non-empty args list')
```

OR

``` python
def __init__(self, args, callback=None):
    if not args:
        args = ['ansible test program']
```

samdoran replied:
Sounds good. I'll sanity check in the `CLI` constructor.

samdoran commented on lib/ansible/cli/arguments/option_helpers.py:
```diff
@@ -178,13 +178,14 @@ def version(prog=None):
 #
 
-def create_base_parser(usage="", desc=None, epilog=None):
+def create_base_parser(usage="", desc=None, epilog=None, prog=None):
```
@abadger I'm not sure in `prog=None` is a good default here.

abadger replied:
Yeah, prog isn't really optional anymore.  I'd probably change the function signature to:
``` python
def create_base_parser(prog, usage="", desc=None, epilog=None):
```

and update the places in ```lib/ansible/cli/*``` that call the function.

samdoran replied:
Done.

samdoran committed:
Require list of args for CLI object, require prog

Add sanity test on CLI constructor to ensure args is defined.
Make prog a required parameter of create_base_parser() and update all uses to pass in the newly required paramater.
Update unit tests to reflect and test these changes.

samdoran committed:
Fix a few more places where an empty list is no longer acceptable

samdoran commented on docs/bin/generate_man.py:
```diff
@@ -116,7 +116,7 @@ def opts_docs(cli_class_name, cli_module_name):
     cli_klass = getattr(__import__("ansible.cli.%s" % cli_module_name,
                                    fromlist=[cli_class_name]), cli_class_name)
-    cli = cli_klass([])
+    cli = cli_klass(['opts_docs'])
```
@abadger I'm not sure if this will effect the docs build.

abadger replied:
@samdoran It probably won't but you could do this:
``` python
cli = cli_klass([cli_name])
```

which is probably "more right" ;-)

abadger replied:
This line is likely not needed with your code and appropriately setting cli as well: https://github.com/ansible/ansible/pull/55728/files/d6b433e03bb2648077cd83ae9352011c735afbc3..4fb108769a560f8f0ea9b6c33854884c2bb12120#diff-8c656f386b78148720508ca0dab12b82R127

samdoran committed:
Few tweaks to genaret_man
